cranelift-codegen/src/binemit/stackmap.rs

```diff
@@ -0,0 +1,124 @@
+use crate::bitset::BitSet;
+use crate::ir;
+use crate::isa::TargetIsa;
+use std::vec::Vec;
+
+/// Wrapper class for longer bit vectors that cannot be represented by a single BitSet
+#[derive(Clone, Debug)]
+pub struct Stackmap {
+    bitmap: Vec<BitSet<u32>>,
+}
+
+impl Stackmap {
+    /// Create a stackmap based on where references are located on a function's stack
+    pub fn from_values(
+        args: &[ir::entities::Value],
+        func: &ir::Function,
+        isa: &dyn TargetIsa,
+    ) -> Self {
+        let stack = &func.stack_slots;
+        let loc = &func.locations;
+        let mut live_ref_in_stack_slot: std::collections::HashSet<ir::StackSlot> =
+            std::collections::HashSet::new();
+        // References can be in registers, but live registers values are pushed onto the stack before calls and traps.
+        for val in args {
+            if let Some(value_loc) = loc.get(*val) {
+                match *value_loc {
+                    ir::ValueLoc::Stack(stack_slot) => live_ref_in_stack_slot.insert(stack_slot),
+                    _ => false,
+                };
+            }
+        }
+
+        // SpiderMonkey stackmap structure:
+        // <trap reg dump> + <general spill> + <frame> + <inbound args>
+        // Bit vector goes from lower addresses to higher addresses
+
+        // TODO: Get trap register layout from Spidermonkey and prepend to bitvector below.
```

Comment:
We had an in-person meeting about what the `Stackmap` metadata format should be at the last all hands. We want to keep both Spidermonkey and Cranelift modular and minimize the inter-language communication. In the end, I believe @julian-seward1 suggested to pass the trap register dump layout format from SpiderMonkey to Cranelift and to prepend. That way, when we pass the metadata back to Spidermonkey, it's just a bitvector. I believe the rationale was to imitate what we did with Ion/Baldr to keep it consistent. 

@sunfishcode and I discussed the plan for this change in person. In order to have working reference types in Cranelift, we need to make changes in Spidermonkey. Instead of trying to do two patches at the same time, we've decided to have the Spidermonkey patch come in another time and get the first steps towards having reference types in Cranelift merged first. 

Iirc, Dan's plan is to have someone else (who is familiar with Spidermonkey) pick up the remaining work. I left the TODO comment there as a reminder of that this change is incomplete and what remains to be done. I can remove the TODO comment if we don't want it?

Reply:
Sounds good to me for the plan; we can keep the comment here, even if I think we shouldn't mimic Spidermonkey's bit vector layout here, and have something higher-level for Cranelift's purpose.